### PR TITLE
Handle invalid BroadcastTo shape in C backend

### LINKED ISSUE #152: BUG: `broadcast_to` crashes when target shape is invalid
### Describe the issue:

`pytensor.tensor.random.utils.params_broadcast_shapes` says that it performs numpy broadcasting on shape tuples. This works kind of fine, unless the shapes are not broadcastable with each other. In those cases, `params_broadcast_shapes` will happily return a shape that is the maximum between the input shape tuples.

This causes a set of problems down the line.
1. `pytensor.tensor.random.utils.params_broadcast_shapes` is used by default in `RandomVariable.infer_shape`. This causes potentially wrong shape inferences, that might raise an exception at runtime.
2. If the output shape returned by `pytensor.tensor.random.utils.params_broadcast_shapes` is used to `broadcast_to`, the resulting tensors will **look like they would work fine**, but they can lead to segfaults or kernel crashes at runtime.

### Reproducable code example:

```python
import pytensor
from pytensor import tensor as pt
import numpy as np

a = pt.as_tensor_variable(np.zeros((3, 2)), name="a")
b = pt.as_tensor_variable(np.ones((1, 4)), name="b")
x = pt.random.normal(name="x", loc=a, scale=b)
x.shape.eval()  # Prints array([3, 4])
x.eval().shape  # ValueError: shape mismatch: objects cannot be broadcast to a single shape.  Mismatch is between arg 0 with shape (3, 2) and arg 1 with shape (1, 4).


from pytensor.tensor.random.utils import params_broadcast_shapes
a_shape, b_shape = params_broadcast_shapes([a.shape, b.shape], ndims_params=[0, 0])
y = pt.random.normal(
    name="y",
    loc=pt.broadcast_to(a, a_shape),
    scale=pt.broadcast_to(b, b_shape),
)
y.shape.eval()  # prints array([3, 4])
y.eval().shape  # Produces a segmentation fault (core dumped)
```

### Error message:

```shell
<details>
<summary>x.eval().shape leads to this error</summary>
ValueError: shape mismatch: objects cannot be broadcast to a single shape.  Mismatch is between arg 0 with shape (3, 2) and arg 1 with shape (1, 4).
Apply node that caused the error: normal_rv{0, (0, 0), floatX, False}(RandomGeneratorSharedVariable(<Generator(PCG64) at 0x7F0B6A29C740>), TensorConstant{[]}, TensorConstant{11}, a{(3, 2) of 0.0}, b{(1, 4) of 1.0})
Toposort index: 0
Inputs types: [RandomGeneratorType, TensorType(int64, (0,)), TensorType(int64, ()), TensorType(float64, (3, 2)), TensorType(float64, (1, 4))]
Inputs shapes: ['No shapes', (0,), (), (3, 2), (1, 4)]
Inputs strides: ['No strides', (0,), (), (16, 8), (32, 8)]
Inputs values: [Generator(PCG64) at 0x7F0B6A29C740, array([], dtype=int64), array(11), 'not shown', array([[1., 1., 1., 1.]])]
Outputs clients: [[], ['output']]

Backtrace when the node is created (use PyTensor flag traceback__limit=N to make it longer):
  File "/home/lpaz/anaconda3/lib/python3.9/site-packages/IPython/terminal/interactiveshell.py", line 678, in interact
    self.run_cell(code, store_history=True)
  File "/home/lpaz/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 2940, in run_cell
    result = self._run_cell(
  File "/home/lpaz/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 2995, in _run_cell
    return runner(coro)
  File "/home/lpaz/anaconda3/lib/python3.9/site-packages/IPython/core/async_helpers.py", line 129, in _pseudo_sync_runner
    coro.send(None)
  File "/home/lpaz/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3194, in run_cell_async
    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
  File "/home/lpaz/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3373, in run_ast_nodes
    if await self.run_code(code, result, async_=asy):
  File "/home/lpaz/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3433, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-16-840c5bd8ff9e>", line 1, in <module>
    x = pt.random.normal(name="x", loc=a, scale=b)

HINT: Use the PyTensor flag `exception_verbosity=high` for a debug print-out and storage map footprint of this Apply node.
</details>

<details>
<summary>y.eval().shape leads to this error</summary>
Segmentation fault (core dumped)
</details>
```

### PyTensor version information:

<details>
Installed via pip
pytensor version is 2.8.11
</details>

### Context for the issue:

_No response_

### PR DESCRIPTION
**Pull request by ricardoV94**
Closes #152

### CODE REVIEW

**Review by lucianopaz — CHANGES_REQUESTED** [review]
Looks good @ricardoV94. I left three comments

File: pytensor/tensor/extra_ops.py [hunk 1]
```diff
@@ -1643,6 +1643,9 @@ def make_node(self, a, *shape):
 
         shape, static_shape = at.infer_static_shape(shape)
 
+        if len(shape) < a.ndim:
+            raise ValueError("Broadcast shape cannot be shorter than input shape")
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py:1647]: Updated

File: tests/tensor/test_extra_ops.py [hunk 2]
```diff
@@ -1253,30 +1253,39 @@ def test_avoid_useless_subtensors(self):
...
+            bcast_in = bcast_fn.vm.storage_map[a]
+            bcast_out = bcast_fn.vm.storage_map[bcast_var]
             assert np.shares_memory(bcast_out[0], bcast_in[0])
 
+    def test_make_node_error_handling(self):
+        with pytest.raises(
+            ValueError, match="Broadcast shape cannot be shorter than input shape"
+        ):
+            broadcast_to(at.zeros((3, 4)), (5,))
+
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply tests/tensor/test_extra_ops.py:1288]: Added a parametrization to that test with invalid values. It seems to be behaving fine

File: pytensor/tensor/extra_ops.py [hunk 3]
```diff
@@ -1715,6 +1718,23 @@ def c_code(self, node, name, inputs, outputs, sub):
...
+            {
+                if ((PyArray_DIMS(%(x)s)[i] != 1) && (PyArray_DIMS(%(x)s)[i] != itershape[i + offset]))
+                {
+                    PyErr_Format(PyExc_ValueError,
+                                 "Shape mismatch in broadcast_to: target shape[%%i] = %%lld is incompatible with input shape = %%lld.",
+                                 i,
+                                 (long long int) itershape[i + offset],
+                                 (long long int) PyArray_DIMS(%(x)s)[i]
+                    );
+                    %(fail)s
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply pytensor/tensor/extra_ops.py:1732]: I move the creation of `iter` to after the check, even though the new test param did not show anything wrong

File: tests/tensor/test_extra_ops.py [hunk 4]
```diff
@@ -1313,7 +1333,7 @@ def test_memory_leak(self):
             blocks_last = blocks_i
 
         tracemalloc.stop()
-        assert np.allclose(np.mean(block_diffs), 0)
+        assert np.mean(block_diffs) <= (0 + 1e-8)
```
Comment (ricardoV94) [inline tests/tensor/test_extra_ops.py:1336]: This was failing with negative values, which should not be problematic

**Review by lucianopaz — APPROVED** [review]
LGTM @ricardoV94. I don't know what do make out of the performance regression. I doesn't seem related to this PR but I can't seem to find a way to get the detailed benchmark results from the cache or from this PR.

**Comment by ricardoV94** [comment]
The last cached benchmark might have been a CI fluke and this one looks bad in comparison?

Only the C scan test could possibly have been affected so I'll run it locally before and after the changes to see if it complains.

**Comment by lucianopaz** [comment]
Looks like the performance regression problems are happening because the tests that use random are not being seeded, and different seeds can lead to very different run times

**Comment by ricardoV94** [comment]
I'll merge this and seed the benchmark in another PR. 

I am surprised that the speed of logsumexp is value dependent. Maybe the max is faster sometimes because of specific signs. Or log requires different number of terms depending on magnitude?

**Pull request merged**